x/qgb/orchestrator/orchestrator_client.go

```diff
@@ -0,0 +1,174 @@
+	valsetsChan := make(chan types.Valset, 10)
+
+	results, err := oc.tendermintRPC.Subscribe(
+		ctx,
+		"valset-changes",
+		fmt.Sprintf("%s.%s='%s'", types.EventTypeValsetRequest, sdk.AttributeKeyModule, types.ModuleName),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+	nonces := make([]uint64, 10000)
+
+	go func() {
+		defer close(valsetsChan)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-results:
+				valsets, err := oc.querier.QueryLastValsets(ctx)
+				if err != nil {
+					oc.logger.Error(err.Error())
+					return
+				}
+
+				// todo: double check that the first validator set is found
+				if len(valsets) < 1 {
+					oc.logger.Error("no validator sets found")
+					return
+				}
+
+				valset := valsets[0]
+
+				// Checking if we already signed this valset
+				resp, err := oc.querier.QueryValsetConfirm(ctx, valset.Nonce, oc.orchestratorAddress)
+				if err != nil {
+					oc.logger.Error(err.Error())
+					return
+				}
+				if resp == nil && !contains(nonces, valset.Nonce) {
+					nonces = append(nonces, valset.Nonce)
+					valsetsChan <- valset
+				}
+			}
+		}
+	}()
+
+	return valsetsChan, nil
+}
+
+func (oc *orchestratorClient) SubscribeDataCommitment(ctx context.Context) (<-chan ExtendedDataCommitment, error) {
+	dataCommitments := make(chan ExtendedDataCommitment)
+
+	// queryClient := types.NewQueryClient(orchestratorClient.qgbRPC)
+
+	// resp, err := queryClient.Params(ctx, &types.QueryParamsRequest{})
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	// params := resp.Params
+	q := coretypes.EventQueryNewBlockHeader.String()
+	results, err := oc.tendermintRPC.Subscribe(ctx, "height", q)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer close(dataCommitments)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case ev := <-results:
+				eventDataHeader := ev.Data.(coretypes.EventDataNewBlockHeader)
+				height := eventDataHeader.Header.Height
+				// todo: refactor to ensure that no ranges of blocks are missed if the
+				// parameters are changed
+				if height%int64(types.DataCommitmentWindow) != 0 {
+					continue
+				}
+
+				// TODO: calculate start height some other way that can handle changes
+				// in the data window param
+				startHeight := height - int64(types.DataCommitmentWindow)
+				endHeight := height
+
+				// create and send the data commitment
+				dcResp, err := oc.tendermintRPC.DataCommitment(
+					ctx,
+					fmt.Sprintf("block.height >= %d AND block.height <= %d",
+						startHeight,
+						endHeight,
+					),
+				)
+				if err != nil {
+					oc.logger.Error(err.Error())
+					continue
+				}
+
+				// TODO: store the nonce in the state somewhere, so that we don't have
+				// to assume what the nonce is
+				nonce := uint64(height) / types.DataCommitmentWindow
+
+				dataCommitments <- ExtendedDataCommitment{
+					Commitment: dcResp.DataCommitment,
+					Start:      startHeight,
+					End:        endHeight,
+					Nonce:      nonce,
+				}
```

Comment:
without any delay here, I think we will quickly fill this channel

Reply:
That depends on the data window and the block time, I assume. If the data window is 100 blocks and block time is 15sec, then it wouldn't fill that fast. If it is 2 blocks and block time of 5sec then yes it will,  and we will risk of losing data commitments. 
However, this only stands for the current design. But, when we will have data commitment requests, I guess it will be better.